doc/layer2-forwarding-enhancements/SONiC Layer 2 Forwarding Enhancements HLD.md

```diff
@@ -0,0 +1,444 @@
+
+
+The internal FDB data structure  will be enhanced to store the FDB type (static/dynamic). Currently, FDB is stored as c++ set. It will be changed to c++ map to store (key, value), where key is (MAC, bv_id) and value is FDB type.
+FDB type is needed to add Static FDB to saved FDB after flush.
+
+
+	**FDB Data structure in Orchagent:**
+
+
+<img src="images/fdbDataStructure.jpg" width="800" height="500">
+
+In current Sonic implementation, when a FDB event is received with VLAN SAI object ID, the VLAN ID is retrieved by traversing through all the ports and VLANs in the orchagent DB and finding the one that matches the object ID. Also, the Port details are retrieved by traversing all the port and VLANs. This approach is inefficient as it may traverse through the number of VLANs configured for each FDB learn event. Worst case time is O(M*N) where M is the number of MAC events and N is the number of entries in port data structure.
+
+
+A new  unordered_map containing a mapping between SAI object ID (can be either Port object ID or Bridge Port object ID or VLAN object ID) and Name (port or VLAN alias) will be added which can be looked up in O(1) time. The Port or VLAN alias retrieved from this will be looked up in the Port data structure.
+
+
+	**Object ID to alias mapping.**
+
+
+<img src="images/oidToName.jpg" width="800" height="500">
+
+
+**Changes for FDB flush.**
+When a FDB flush request is received, Orchagent will send a bulk flush request to syncd. Individual FDB delete response from SAI will be used to delete FDB entries in ASIC_DB, STATE_DB and orchagent data structure. The delete response will be received as AGED event to syncd, which will delete the entry in ASIC_DB and notify Orchagent.
+
+
+When a port operational status goes down, portsorch will call fdborch API to trigger FDB flush. Only dynamic FDB will be flushed.
+
+
+When a port is removed from VLAN, portsorch will call fdborch API to trigger FDB flush. Both static and dynamic FDB will be flushed. Also, the static FDB will be added to the temporary DB in orchagent when delete response is received, so that it can programmed back when port is added back to VLAN.
+
+
+When Spanning tree state changes, protocol component within orchagent will call fdborch API to flush FDB by either per Port or per Port per VLAN. Only dynamic FDB will be flushed.
+
+
+	**FDB Flush due to Port operational state down:**
+
+
+<img src="images/portDownFlush.jpg" width="800" height="500">
+
+
+	**FDB Flush due to Port removal from VLAN:**
+
+
+<img src="images/portVlanRemoveFlush.jpg" width="800" height="500">
+
+
+	**FDB Flush due to spanning tree state change:**
+
+
+<img src="images/stpFlush.jpg" width="800" height="500">
+
+
+**Changes for handling MAC move events.**
+Hardware can generate a single MAC move event instead of generating 2 events (del-old mac and add-new mac). Orchagent code changes will be done to replace the existing FDB entry with the new port in STATE_DB. 
```

Comment:
Do you plan to add a new code branch for MAC move event or obsolete the current 2 events way? I suggest to add a new Move event handling and also keep the current ones.

Reply:
Some platforms generate 2 events (age and learn) for mac move. It will continue to be supported. Some platforms generate only 1 move event. This document talks about adding support for this. Existing behavior will not change.